vllm/v1/sample/rejection_sampler.py

```diff
@@ -118,54 +193,146 @@ def flashinfer_sample(
+
+            output_token_ids = target_token_ids_tensor
+            output_token_ids[~generate_mask] = INVALID_TOKEN_ID
         else:
-            raise NotImplementedError(
-                "Currently, only greedy sampling is supported by "
-                "rejection sampler.")
-        # Identify valid positions (non-padding).
-        valid_mask = output_token_ids != INVALID_TOKEN_ID
-        # Generate mask with bonus token.
-        generate_mask = torch.cat([
-            accept_mask,
-            torch.zeros(accept_mask.size(0), 1, device=accept_mask.device)
-        ],
-                                  dim=1).to(torch.bool) & valid_mask
-        zeros_mask = (generate_mask == 0)
-        first_zero_idx = zeros_mask.float().argmax(dim=1)
-        # Figure out which rows actually contain at least one zero.
-        rows_with_zero = zeros_mask.any(dim=1)
-        # Use indexing to set the first zero in each of those rows to 1.
-        generate_mask[rows_with_zero, first_zero_idx[rows_with_zero]] = 1
-
-        output_token_ids[~generate_mask] = INVALID_TOKEN_ID
+            # Reference: https://arxiv.org/pdf/2211.17192
+            # 1. Extract the probabilities of the draft tokens.
+            # [batch_size, max_spec_len]
+            batch_size = draft_token_ids_tensor.size(0)
+            max_spec_len = draft_token_ids_tensor.size(1)
+            invalid_idx = draft_token_ids_tensor == INVALID_TOKEN_ID
+            draft_token_ids_tensor[invalid_idx] = 0
+            assert draft_probs is not None
+            draft_token_probs = draft_probs.gather(
+                dim=-1, index=draft_token_ids_tensor.unsqueeze(-1)).squeeze(-1)
+            target_token_probs = target_probs.gather(
+                dim=-1, index=draft_token_ids_tensor.unsqueeze(-1)).squeeze(-1)
+            # Force the probabilities of invalid tokens to inf
+            # so that they are not accepted.
+            draft_token_probs[invalid_idx] = float('inf')
+
+            # 2. Generate uniform samples.
+            # [batch_size, max_spec_len + 1]
+            uniform_samples = _create_uniform_samples(
+                sampling_metadata.generators, batch_size, max_spec_len,
+                target_probs.device)
+
+            # 3. Accept or reject the samples.
+            # [batch_size, max_spec_len]
+            accepted = uniform_samples <= target_token_probs / draft_token_probs
+            accept_mask = accepted.cumprod(dim=1)
+            # Set the token ids to the draft token ids if accepted, otherwise
+            # set them to INVALID_TOKEN_ID.
+            accepted_token_ids = (draft_token_ids_tensor * accept_mask +
+                                  INVALID_TOKEN_ID * (1 - accept_mask))
+
+            # 4. Adjust the distribution for the recovered tokens.
+            bonus_prob = torch.clamp(target_probs[:, :-1, :] - draft_probs,
+                                     min=1e-5)
```

Comment:
As we discussed offline, please use smaller values like 1e-8

Reply:
I changed it to `torch.finfo(torch.float32).tiny`, which is what we did in v0